presto-main/src/main/java/com/facebook/presto/cost/ComparisonStatsCalculator.java

```diff
@@ -221,25 +221,25 @@ private ComparisonStatsCalculator() {}
         PlanNodeStatsEstimate inputNullsFiltered = inputStatistics.mapOutputRowCount(size -> size * nullsFilterFactor);
         SymbolStatsEstimate leftNullsFiltered = leftExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0);
         SymbolStatsEstimate rightNullsFiltered = rightExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0);
-        Optional<PlanNodeStatsEstimate> equalityStats = estimateExpressionEqualToExpression(
+        PlanNodeStatsEstimate equalityStats = estimateExpressionEqualToExpression(
                 inputNullsFiltered,
                 leftNullsFiltered,
                 leftExpressionSymbol,
                 rightNullsFiltered,
                 rightExpressionSymbol);
-        if (!equalityStats.isPresent()) {
-            return Optional.empty();
+        if (equalityStats.isOutputRowCountUnknown()) {
```

Comment:
I am against relying on implicit propagation of NaNs. For me, it makes reading & reasoning about the code harder. 
This is because NaNs are not "totally contagious". When inputs are NaNs, the following code will still return 1 (would you expect 1 or NaN??):
```
if (ndv < rowCount) {
  return ndv / rowCount;
} else {
  return 1;
}
```

Reply:
Yeah, i think so. Anyhow, if the row count is unknown, it doesn't make much sense to propagate some single symbol stats.

Reply:
I generally agree. I can imagine one possible use of propagating some single symbols stats in troubleshooting -- especially if we make EXPLAIN (optionally) show all the column stats.

Reply:
I think one just needs to remember that we have kind of three value logic here. We could wrap it into dedicated syntax, e.g:
```
double value = compare(left).greaterThan(right)
   .onTrue(x)
   .onFalse(y)
   .onNaN(z)                        
```
but I don't think it's work extra bolerplate.